### PR TITLE
Disable local accounts for aks aad clusters

### LINKED ISSUE #2129: Support disableLocalAccounts for AKS clusters
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
Add support for setting [DisableLocalAccounts](https://docs.microsoft.com/en-us/azure/aks/managed-aad#disable-local-accounts). This is a security feature that prevents users from fetching the admin kubeconfig which has a 2 year certificate with system:masters privileges.

In order to support this, CAPZ will need to be changed to fetch user kubeconfig instead of admin kubeconfig in the [GetCredentials](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/c444fdb66713b580ebc65ea8ce4436715358bf78/azure/services/managedclusters/client.go#L68) function.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]
Note that this is only relevant for AKS clusters

### PR DESCRIPTION
**Pull request by LochanRn**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Support to disable local accounts for aks aad clusters

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2129

**Special notes for your reviewer**:

As access for admin kubeconfig will be disabled, the user must ensure to add the service principle to the admin group configured for enabling AAD.

- [ ] cherry-pick candidate

**TODOs**:

- [x] squashed commits
- [x] includes documentation
- [X] adds unit tests

**Release note**:

```release-note
Support disabling of local accounts for AKS clusters with Azure Active Directory enabled. Disables the back door to get admin kubeconfig for AAD based clusters. The user must ensure to add the service principle to admin groups in azure AAD for capz to be able to deploy the target cluster..
```

### CODE REVIEW

File: azure/services/managedclusters/managedclusters.go
Comment (Jont828) [inline azure/services/managedclusters/managedclusters.go:50]:
```suggestion
	GetAdminKubeconfigData() []byte
```
For consistency of the capitalization through the repo

File: azure/services/managedclusters/spec.go
Comment (Jont828) [inline azure/services/managedclusters/spec.go:126]:
```suggestion
	// DisableLocalAccounts disables getting static credentials for this cluster when set. Expected to only be used for AAD clusters.
```

File: api/v1beta1/azuremanagedcontrolplane_types.go
Comment (Jont828) [inline api/v1beta1/azuremanagedcontrolplane_types.go:217]:
```suggestion
	// DisableLocalAccounts disables getting static credentials for this cluster when set. Expected to only be used for AAD clusters.
```

File: api/v1beta1/azuremanagedcontrolplane_types.go
Comment (Jont828) [inline api/v1beta1/azuremanagedcontrolplane_types.go:219]: Is there a difference between how we handle nil and false? I'm thinking we could just make it a normal bool to avoid unecessary pointers.

**Review by Jont828 — COMMENTED** [review]
Nice work! Just some small changes

File: api/v1beta1/azuremanagedcontrolplane_types.go
Comment (LochanRn) [reply api/v1beta1/azuremanagedcontrolplane_types.go:219]: In the entire spec all optional fields are generally pointers :)

File: azure/scope/managedcontrolplane.go
Comment (LochanRn) [reply azure/scope/managedcontrolplane.go:673]: Hmm what is the change that is suggested here 😅

Not sure because the change suggested and the existing function name both are same. 
SetAdminKubeconfigData(kubeconfigData

File: api/v1beta1/azuremanagedcontrolplane_types.go
Comment (LochanRn) [reply api/v1beta1/azuremanagedcontrolplane_types.go:217]: I copied the comments from the azure sdk :) 

https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/containerservice/armcontainerservice/v4@v4.3.0#ManagedClusterProperties.DisableLocalAccounts

File: azure/scope/managedcontrolplane.go
Comment (LochanRn) [reply azure/scope/managedcontrolplane.go:477]: ack

File: azure/scope/managedcontrolplane.go
Comment (LochanRn) [reply azure/scope/managedcontrolplane.go:673]: Ohh the c in Kubeconfig nice catch thanks. ack.

File: azure/services/managedclusters/managedclusters.go
Comment (LochanRn) [reply azure/services/managedclusters/managedclusters.go:50]: ack

File: azure/scope/managedcontrolplane.go
Comment (willie-yao) [inline azure/scope/managedcontrolplane.go:672]: nit for consistency with the comment above
```suggestion
// SetAdminKubeconfigData sets admin kubeconfig data.
```

File: azure/services/managedclusters/managedclusters.go
Comment (willie-yao) [inline azure/services/managedclusters/managedclusters.go:39]: Just for my own understanding: is this supposed to be hard-coded?

File: azure/services/managedclusters/managedclusters.go
Comment (willie-yao) [inline azure/services/managedclusters/managedclusters.go:170]: Is it possible to add a unit test for this function? I think it can be useful here but if it requires too much mocking we can track this in a separate issue.

File: azure/services/managedclusters/managedclusters.go
Comment (LochanRn) [reply azure/services/managedclusters/managedclusters.go:39]: yes

https://azure.github.io/kubelogin/concepts/aks.html?highlight=6dae42f8-4368-4678-94ff-3960e28e3630#azure-kubernetes-service-aad-server

File: azure/services/managedclusters/managedclusters.go
Comment (LochanRn) [reply azure/services/managedclusters/managedclusters.go:170]: let me try to do that.

**Comment by LochanRn** [comment]
`Error: accumulating resources: accumulation err='accumulating resources from 'https://raw.githubusercontent.com/Azure/azure-container-networking/v1.5.3/hack/manifests/cni-installer-v1.yaml': Get "https://raw.githubusercontent.com/Azure/azure-container-networking/v1.5.3/hack/manifests/cni-installer-v1.yaml": dial tcp [2405:200:1607:2820:41::36]:443: i/o timeout': hit 27s timeout running '/usr/bin/git fetch --depth=1 origin HEAD'
make[1]: *** [generate-addons] Error 1
make: *** [generate] Error 2`

Facing this issue. Any help will be much appreciated. :)

File: controllers/azuremanagedcontrolplane_reconciler.go
Comment (CecileRobertMichon) [inline controllers/azuremanagedcontrolplane_reconciler.go:141]: @LochanRn fyi this change conflicts with #4052 so whichever PR doesn't merge first will have to do some rebasing

Comment (LochanRn) [reply controllers/azuremanagedcontrolplane_reconciler.go:141]: @CecileRobertMichon  okay seems like a big piece, and a great one !!

**Comment by willie-yao** [comment]
PR looks good pending the unit test for `ReconcileKubeconfig`! Although I don't think it would be blocking the PR for merge.

/lgtm

File: azure/services/managedclusters/managedclusters.go
Comment (CecileRobertMichon) [reply azure/services/managedclusters/managedclusters.go:39]: Can you please add a comment with the link for future reference?

File: api/v1beta1/azuremanagedcontrolplane_webhook.go
Comment (CecileRobertMichon) [inline api/v1beta1/azuremanagedcontrolplane_webhook.go:625]: can we use the `webhookutils.ValidateImmutable` function for this?

File: azure/services/managedclusters/managedclusters.go
Comment (LochanRn) [reply azure/services/managedclusters/managedclusters.go:39]: ack

File: azure/scope/managedcontrolplane.go
Comment (CecileRobertMichon) [inline azure/scope/managedcontrolplane.go:464]: nit: `AreLocalAccountsDisabled` is more gramatically correct

File: azure/services/managedclusters/managedclusters.go
Comment (LochanRn) [reply azure/services/managedclusters/managedclusters.go:170]: @willie-yao I am unable to figure out how to mock this and add more testing for this, actually a part of the testing is already covered in the below test-case. 

https://github.com/spectrocloud/cluster-api-provider-azure/blob/0abf40eced4dac96e0869c036df5cd243bfd3e49/azure/services/managedclusters/managedclusters_test.go#L105-#L142

What else do you think i can do to improve the coverage ?

File: azure/services/token/client.go
Comment (CecileRobertMichon) [inline azure/services/token/client.go:10]: azureautorest is deprecated and we're actively working toward getting rid of it (see #2974), is it possible to not introduce new usage? Any alternative way to do what you're doing without using azureautorest libraries?

cc @mboersma

File: azure/services/token/client.go
Comment (CecileRobertMichon) [inline azure/services/token/client.go:57]: we shouldn't be defaulting the Azure environment inside the clients, we have defaulting in place already in the controller, defaulting in two different places could result in future bugs if the default changes

File: api/v1beta1/azuremanagedcontrolplane_webhook.go
Comment (LochanRn) [reply api/v1beta1/azuremanagedcontrolplane_webhook.go:625]: I guess no because we can disableLocalAccounts on a running cluster but cannot enable it back after that, so if it is nil it can be updated to true, but once updated to true it cannot be set to false or made nil.

File: azure/services/managedclusters/managedclusters.go
Comment (willie-yao) [reply azure/services/managedclusters/managedclusters.go:170]: I think in this case, the test coverage is fine as-is since it is already covered in the test case listed above. Thanks for looking into it!

File: azure/services/token/client.go
Comment (LochanRn) [reply azure/services/token/client.go:10]: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/1a70b0108b4049f08e9918a56123233e6a586f91/azure/scope/clients.go#L45

This function returns the name instead of env, which is why i had to recalculate the env. I will update the function to return just the env. Then i don't have to figure out the cloud env all over again.

**Comment by Jont828** [comment]
LGTM aside from pending comments and once tests pass

File: azure/services/token/client.go
Comment (CecileRobertMichon) [reply azure/services/token/client.go:26]: This is still introducing a new use of go-autorest which is deprected. I think you should be able to use `azure.ARMClientOptions` (see https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/services/publicips/client.go#L37 for example usage) to avoid constructing your own ` azcore.ClientOptions` altogether

This will require reverting the last commit to go back since we no longer need to return the full env.

Comment (LochanRn) [reply azure/services/token/client.go]: done :white_check_mark:

**Review by CecileRobertMichon — APPROVED** [review]
/lgtm
/approve

Thank you @LochanRn!

@nojnhuh FYI this PR might be disruptive for both #4052 and #4069 but it's worth merging as-is and rebasing the two big PRs on top since it's a feature that's been asked for for a while

**Comment by LochanRn** [comment]
@CecileRobertMichon PTAL i added docs !

**Review by CecileRobertMichon — COMMENTED** [review]
minor nits

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm

File: azure/services/managedclusters/managedclusters.go
Comment (nojnhuh) [inline azure/services/managedclusters/managedclusters.go:170]: Is this comment correct or should it be referring to the user kubeconfig instead of the admin one?

File: azure/services/managedclusters/managedclusters.go
Comment (nojnhuh) [reply azure/services/managedclusters/managedclusters.go:211]: I don't see any major blockers with ASO. It might be clunky to add this token to the user kubeconfig if ASO doesn't do that for us, but I don't think that would be any more involved than this already is.

@LochanRn Would you be able to help test this feature in #4069 when I resolve conflicts there?

File: docs/book/src/topics/managedcluster.md
Comment (nojnhuh) [inline docs/book/src/topics/managedcluster.md:294]: I'm struggling to set this up. Is there a link to Azure docs that describe how to do this?

File: azure/services/managedclusters/managedclusters.go
Comment (LochanRn) [reply azure/services/managedclusters/managedclusters.go:170]: This comment is correct, if we refer to the user kubeconfig we will need kubelogin to access the cluster.

File: docs/book/src/topics/managedcluster.md
Comment (LochanRn) [reply docs/book/src/topics/managedcluster.md:294]: Can we get into a call sometime today ? I can explain how this works.

Comment (nojnhuh) [reply docs/book/src/topics/managedcluster.md:294]: Synced with @LochanRn and it looks like I need to create a group with a `type` of "Security" to be able to add a Service Principal as a member and I don't seem to have permission to do that.

@CecileRobertMichon Is there a chance I could be granted that permission to test this? Not a huge deal if not since @LochanRn agreed to help test this on my ASO PR but if I could test myself that would be a lot easier.

**Pull request merged**
